### PR TITLE
Single entry wal

### PR DESCRIPTION
**Pull request by hackaugusto**
Introduces the WAL class. This is the entry point of state changes for the app.

The `PickleSerializer` and `SQLiteStorage` classes are almost identitical, the biggest change is the breaking in the `SQLiteStorage` interface (get methods were renamed and return values changed).

### CODE REVIEW

**Comment by hackaugusto** [comment]
Ideally `get_events_by_block` and `get_events_by_identifier` would not expose the internal database identifiers, I didn't remove it yet because some tests seem to rely on it.

File: raiden/storage/sqlite.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
+# -*- coding: utf-8 -*-
+import sqlite3
+import threading
+
+
+def sqlite_sanity_check(conn):
```
Comment (palango) [inline raiden/storage/sqlite.py]: This seems to be unnecessary for python 3.6 which is required by now.

File: raiden/storage/sqlite.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+                'INSERT INTO state_changes(identifier, data) VALUES(null, ?)',
+                (serialized_data,)
+            )
+            last_id = cursor.lastrowid
+
+        return last_id
+
+    def write_state_snapshot(self, statechange_id, snapshot):
+        # TODO: Snapshotting is not yet implemented. This is just skeleton code
+        # Issue: https://github.com/raiden-network/raiden/issues/593
```
Comment (palango) [inline raiden/storage/sqlite.py]: this issue is closed: https://github.com/raiden-network/raiden/issues/593
what's the status of this todo?

File: raiden/storage/sqlite.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+            )
+            last_id = cursor.lastrowid
+
+        return last_id
+
+    def write_events(self, state_change_id, block_number, events):
+        """ Save events.
+
+        Args:
+            events_data (List[Tuple]): Must be a list of tuples of the form:
```
Comment (palango) [inline raiden/storage/sqlite.py]: this is done in the function body now.

File: raiden/storage/serialize.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+class PickleSerializer:
+    @staticmethod
+    def serialize(transaction):
+        # Some of our StateChange classes have __slots__ without having a __getstate__
+        # As seen in the SO question below:
+        # http://stackoverflow.com/questions/2204155/why-am-i-getting-an-error-about-my-class-defining-slots-when-trying-to-pickl#2204702
+        # We can either add a __getstate__ to all of them or use the `-1` protocol and be
+        # incompatible with ancient python version. Here I opt for the latter.
+        return pickle.dumps(transaction, -1)
```
Comment (palango) [inline raiden/storage/serialize.py]: Should we explicitly state here protocol version 3 or 4? -1 might give some problems with new protocols in the future. anyways, not a big deal.

File: raiden/storage/sqlite.py (hunk 2 above)
Comment (hackaugusto) [reply raiden/storage/sqlite.py]: https://github.com/raiden-network/raiden/issues/682

File: raiden/storage/wal.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+from raiden.transfer.architecture import StateManager
+
+InternalEvent = namedtuple(
+    'InternalEvent',
+    ('identifier', 'state_change_id', 'block_number', 'event_object'),
+)
+
+
+class WriteAheadLog:
```
Comment (LefterisJP) [inline raiden/storage/wal.py]: These will eventually replace the old `log_and_dispatch`in the `StateMachineEventHandler` and `StateChangeLog` in `transfer/log.py` I suppose?

File: raiden/storage/sqlite.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
+# -*- coding: utf-8 -*-
+import sqlite3
+import threading
+
+
+class SQLiteStorage:
```
Comment (LefterisJP) [inline raiden/storage/sqlite.py]: So when this class is integrated we can get rid of `StateChangeLogSQLiteBackend` in transfer/log.py?

File: raiden/storage/wal.py (hunk 5 above)
Comment (hackaugusto) [reply raiden/storage/wal.py]: yes

File: raiden/storage/sqlite.py (hunk 6 above)
Comment (hackaugusto) [reply raiden/storage/sqlite.py]: yes

**Pull request merged**